cdap-data-fabric/src/main/java/io/cdap/cdap/data2/metadata/lineage/field/FieldLineageTable.java

```diff
@@ -496,19 +499,27 @@ public List<EndPoint> getEndpoints(String namespaceId, ProgramRunId programRunId
         StructuredRow row = iterator.next();
         ProgramRunId retrievedProgramRunId = GSON.fromJson(row.getString(
           StoreDefinition.FieldLineageStore.PROGRAM_RUN_FIELD), ProgramRunId.class);
-        if (retrievedProgramRunId.equals(programRunId)) {
-          String namespace = row.getString(StoreDefinition.FieldLineageStore.ENDPOINT_NAMESPACE_FIELD);
-          String name = row.getString(StoreDefinition.FieldLineageStore.ENDPOINT_NAME_FIELD);
-          Map<String, String> properties = GSON.fromJson(
-            row.getString(StoreDefinition.FieldLineageStore.ENDPOINT_PROPERTIES_FIELD),
-            new com.google.gson.reflect.TypeToken<Map<String, String>>() { }.getType());
-          EndPoint matchingEndPoint = EndPoint.of(namespace, name, properties);
-          result.add(matchingEndPoint);
+        if (!isProgramRunIdMatchesProgramReferenceAndRunId(retrievedProgramRunId, programReference, runId)) {
+          continue;
         }
+        String namespace = row.getString(StoreDefinition.FieldLineageStore.ENDPOINT_NAMESPACE_FIELD);
+        String name = row.getString(StoreDefinition.FieldLineageStore.ENDPOINT_NAME_FIELD);
+        Map<String, String> properties = GSON.fromJson(
+          row.getString(StoreDefinition.FieldLineageStore.ENDPOINT_PROPERTIES_FIELD), MAP_STRING_TYPE);
+        EndPoint matchingEndPoint = EndPoint.of(namespace, name, properties);
+        result.add(matchingEndPoint);
       }
     }
     return result;
   }
+  private boolean isProgramRunIdMatchesProgramReferenceAndRunId(ProgramRunId programRunId,
+                                                                ProgramReference programReference, String runId) {
+    return programRunId.getProgram().equals(programReference.getProgram()) &&
```

Comment:
instead of this, you can get the ProgramReference from the ProgramRunId as `programRunId.getParent().getProgramReference()` then compare the references directly